commerce_coordinator/apps/lms/views.py

```diff
@@ -401,3 +413,158 @@ def post(self, request):
+        """Calculate the discounted price of the program."""
+        if not program_offer:
+            return program_price
+
+        if program_price < 1:
+            return program_price
+
+        discounted_price = program_price
+        if program_offer.get("discount_type") == CT_ABSOLUTE_DISCOUNT_TYPE:
+            discounted_price -= program_offer.get("discount_value_in_cents")
+        else:
+            discounted_price -= discounted_price * program_offer.get("discount_value_in_cents") / 10000
+
+        if discounted_price < 1:
+            return program_price
+
+        return discounted_price
+
+    def _extract_parent_course_key(self, course_key: str) -> str:
+        """
+        Extract course_id for comparision
+        """
+        match = re.search(r'course-v1:([^+]+)\+([^+]+)', course_key)
+        if match:
+            return f"{match.group(1)}+{match.group(2)}"
+        else:
+            return None
+
+    def _extract_course_key_from_variant_key(self, variant_key: str) -> tuple:
+        """
+        Extract course key for comparision
+        """
+        match = re.match(r"([a-f0-9-]+|[a-zA-Z]+)\+(.+)", variant_key)
+        course_key = match.group(2) if match else ""
+
+        return course_key
+
+    def _filter_enrollable_program_entitlements(
+            self,
+            program_variants: List[dict],
+            user_enrollable_course_keys: List[str],
+            enrollments: List[str],
+            entitlements: List[str],
+    ) -> List[str]:
+        """Filter out bundle entitlements in which user can enroll."""
+
+        filtered_entitlement_skus = []
+        if user_enrollable_course_keys:
+            for variant in program_variants:
+                course_key = self._extract_course_key_from_variant_key(variant.get("variant_key"))
+                if course_key in user_enrollable_course_keys:
+                    entitlement_sku = variant.get("entitlement_sku")
+                    filtered_entitlement_skus.append(entitlement_sku)
+
+            return filtered_entitlement_skus
+
+        purchasable_courses = []
+        for usr_enr in enrollments:
+            if not usr_enr.get("is_active") or usr_enr.get("mode") in EXCLUDED_ENROLLMENT_COURSE_MODES:
+                course_key = usr_enr.get("course_details", {}).get("course_id")
+                purchasable_courses.append(self._extract_parent_course_key(course_key))
+
+        purchasable_entitlements = []
+        for usr_ent in entitlements:
+            if usr_ent.get("expired_at"):
+                purchasable_entitlements.append(usr_ent.get("course_uuid"))
+
+        for variant in program_variants:
+            entitlement_sku = variant.get("entitlement_sku")
+            variant_key = variant.get("variant_key", "")
+            course_key = self._extract_course_key_from_variant_key(variant_key)
+
+            if course_key in purchasable_courses:
```

Comment:
we will have course run keys in purchasbale_courses and we can use string include to see if course_key is in purchasable course_run_keys